tensorstore/kvstore/s3/s3_request_builder.cc

```diff
@@ -0,0 +1,317 @@
+using ::tensorstore::internal::ParseGenericUri;
+using ::tensorstore::internal::ParsedGenericUri;
+using ::tensorstore::internal_storage_s3::IsValidBucketName;
+
+namespace tensorstore {
+namespace internal_storage_s3 {
+
+
+namespace {
+
+// See description of function UriEncode at this URL
+// https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+constexpr AsciiSet kUriUnreservedChars{
+    "abcdefghijklmnopqrstuvwxyz"
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "0123456789"
+    "-._~"};
+
+// NOTE: Only adds "/" to kUriUnreservedChars
+constexpr AsciiSet kUriObjectKeyReservedChars{
+    "abcdefghijklmnopqrstuvwxyz"
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "0123456789"
+    "/-._~"};
+}
+
+
+std::string UriEncode(std::string_view src) {
+    std::string dest;
+    PercentEncodeReserved(src, dest, kUriUnreservedChars);
+    return dest;
+}
+
+std::string UriObjectKeyEncode(std::string_view src) {
+    std::string dest;
+    PercentEncodeReserved(src, dest, kUriObjectKeyReservedChars);
+    return dest;
+}
+
+S3RequestBuilder::S3RequestBuilder(std::string_view method,
+                                   std::string endpoint_url)
+    : query_parameter_separator_("?") {
+  request_.method_ = method;
+  request_.url_ = std::move(endpoint_url);
+}
+
+HttpRequest S3RequestBuilder::BuildRequest(
+    std::string_view aws_access_key,
+    std::string_view aws_secret_access_key,
+    std::string_view aws_region,
+    std::string_view payload_hash,
+    const absl::Time & time) {
+
+
+  auto headers = request_.headers();
+  std::sort(std::begin(headers), std::end(headers));
+  std::sort(std::begin(encoded_queries_), std::end(encoded_queries_));
+  auto canonical_request = CanonicalRequest(request_.url(), request_.method(),
+                                            payload_hash, headers,
+                                            encoded_queries_);
+  auto signing_string = SigningString(canonical_request, aws_region, time);
+  auto signature = Signature(aws_secret_access_key, aws_region, signing_string, time);
+  auto auth_header = AuthorizationHeader(aws_access_key, aws_region, signature, headers, time);
+  request_.headers_.emplace_back(std::move(auth_header));
+  return std::move(request_);
+}
+
+S3RequestBuilder& S3RequestBuilder::AddUserAgentPrefix(std::string_view prefix) {
+  request_.user_agent_ = tensorstore::StrCat(prefix, request_.user_agent_);
+  return *this;
+}
+
+S3RequestBuilder& S3RequestBuilder::AddHeader(std::string header) {
+  request_.headers_.emplace_back(std::move(header));
+  return *this;
+}
+
+S3RequestBuilder& S3RequestBuilder::AddQueryParameter(
+    std::string_view key, std::string_view value) {
+  auto enc_value = UriEncode(value);
+  auto enc_key = UriEncode(key);
+  auto parameter = tensorstore::StrCat(
+      query_parameter_separator_,
+      enc_key, "=", enc_value);
+  query_parameter_separator_ = "&";
+  request_.url_.append(parameter);
+  encoded_queries_.push_back({enc_key, enc_value});
```

Comment:
If you want "encoded_queries_" to be sorted for a canonical hash used later, that could be done in HTTPRequest.

Reply:
Put in https://github.com/google/tensorstore/pull/91/commits/e2f3687eed3856324fe9ab819cb88088fad26864 to extract query params off the url and then sort them. Also added the the AWS4 ListBuckets test case (which uses query parameters).